source/s3_buffer_pool.c

```diff
@@ -0,0 +1,410 @@
+            "aligned.");
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
+
+    struct aws_s3_buffer_pool *buffer_pool = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_buffer_pool));
+
+    AWS_FATAL_ASSERT(buffer_pool != NULL);
+
+    buffer_pool->base_allocator = allocator;
+    buffer_pool->chunk_size = chunk_size;
+    buffer_pool->block_size = chunk_size * s_chunks_per_block;
+    /* Somewhat arbitrary number.
+     * Tries to balance between how many allocations use buffer and buffer space
+     * being wasted. */
+    buffer_pool->primary_size_cutoff = chunk_size * 4;
+    buffer_pool->mem_limit = mem_limit - s_buffer_pool_reserved_mem;
+    int mutex_error = aws_mutex_init(&buffer_pool->mutex);
+    AWS_FATAL_ASSERT(mutex_error == AWS_OP_SUCCESS);
+
+    aws_array_list_init_dynamic(
+        &buffer_pool->blocks, allocator, s_block_list_initial_capacity, sizeof(struct s3_buffer_pool_block));
+
+    return buffer_pool;
+}
+
+void aws_s3_buffer_pool_destroy(struct aws_s3_buffer_pool *buffer_pool) {
+    if (buffer_pool == NULL) {
+        return;
+    }
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        AWS_FATAL_ASSERT(block->alloc_bit_mask == 0 && "Allocator still has outstanding blocks");
+        aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+    }
+
+    aws_array_list_clean_up(&buffer_pool->blocks);
+
+    aws_mutex_clean_up(&buffer_pool->mutex);
+    struct aws_allocator *base = buffer_pool->base_allocator;
+    aws_mem_release(base, buffer_pool);
+}
+
+void s_buffer_pool_trim_synced(struct aws_s3_buffer_pool *buffer_pool) {
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks);) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->alloc_bit_mask == 0) {
+            aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+            aws_array_list_erase(&buffer_pool->blocks, i);
+            /* do not increment since we just released element */
+        } else {
+            ++i;
+        }
+    }
+}
+
+void aws_s3_buffer_pool_trim(struct aws_s3_buffer_pool *buffer_pool) {
+    aws_mutex_lock(&buffer_pool->mutex);
+    s_buffer_pool_trim_synced(buffer_pool);
+    aws_mutex_unlock(&buffer_pool->mutex);
+}
+
+struct aws_s3_buffer_pool_ticket *aws_s3_buffer_pool_reserve(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+    AWS_PRECONDITION(buffer_pool);
+
+    if (buffer_pool->has_reservation_hold) {
+        return NULL;
+    }
+
+    if (size == 0) {
+        AWS_LOGF_ERROR(AWS_LS_S3_CLIENT, "Could not reserve from buffer pool. 0 is not a valid size.");
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
```

Comment:
Real code never checks the error-code from `aws_s3_buffer_pool_reserve()`. Real code assumes it failed due to the memory limit being temporarily exceeded.
- so if size was 0, the meta-request would get stuck perpetually trying to acquire a ticket
- similarly, if `size > mem_limit` the meta-request would be stuck perpetually trying to acquire a ticket. I realize you adjusted the client's `max_part_size` to make this impossible, but it might be worth an ASSERT or FATAL_ASSERT here in case client code changes in the future and something slips through


Anyway, we should probably do one of these things:
1) real code calling aws_s3_buffer_pool_reserve() continues to assume that it can only fail when the mem limit is temporarily exceeded. And the reserve() function here FATAL_ASSERTs that the size is legal (not 0, not exceeding mem_limit)
2) OR real code calling aws_s3_buffer_pool_reserve() checks to differentiate the expected "waiting for memory" from other errors that terminate the meta-request

Reply:
yeah, good point. size == 0 or > mem_lim is really an indication that something wrong went on before, so for simplicity sake, i think fatal assert makes more sense.